src/main/java/org/spongepowered/api/util/command/args/SelectorCommandElement.java

```diff
@@ -0,0 +1,109 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.util.command.args;
+
+import static com.google.common.base.Preconditions.*;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.common.base.Predicates;
+import com.google.common.collect.Collections2;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+
+import org.spongepowered.api.Game;
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.text.Texts;
+import org.spongepowered.api.text.selector.Selector;
+import org.spongepowered.api.text.selector.SelectorType;
+import org.spongepowered.api.text.selector.Selectors;
+import org.spongepowered.api.util.command.CommandSource;
+
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+/**
+ * Abstract command element that matches values based on a {@link Selector}.
+ */
+public abstract class SelectorCommandElement extends PatternMatchingCommandElement {
+    
+    private final Game game;
+    
+    private static final Function<SelectorType, String> SELECTOR_NAME = new Function<SelectorType, String>() {
+
+        @Override
+        public String apply(SelectorType input) {
+            return input.getId();
+        }
+    };
+    
+    protected SelectorCommandElement(@Nullable Text key, Game game) {
+        super(key);
+        this.game = checkNotNull(game);
+    }
+
+    @Nullable
+    @Override
+    protected Object parseValue(CommandSource source, CommandArgs args) throws ArgumentParseException {
+        Object state = args.getState();
+        
+        final String arg = args.next();
+        if (arg.startsWith("@")) {
+            try {
+                return Selectors.parse(arg).resolve(source);
+            } catch (RuntimeException e) {
+                throw args.createError(Texts.of(e.getMessage()));
+            }
+        }
+        
+        args.setState(state);
+        return super.parseValue(source, args);
+    }
+
+    @Override
+    public List<String> complete(CommandSource src, CommandArgs args, CommandContext context) {
+        Iterable<String> choices = getChoices(src);
+        Object state = args.getState();
+        final Optional<String> nextArg = args.nextIfPresent();
+        args.setState(state);
+        
+        if (nextArg.isPresent()) {
+            if (nextArg.get().startsWith("@")) {
+                if (nextArg.get().contains("]")) {
+                    return ImmutableList.of();
+                }
+                if (nextArg.get().contains("[")) {
+                    // TODO tab complete selector arguments?
```

Comment:
Probably just hardcoding, also `@p` supports it too so it may be best just to complete it for all? Not sure if we really need to separate out the different types.